pkg/expr/func.go

```diff
@@ -0,0 +1,242 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expr
+
+import (
+	"reflect"
+	"strings"
+
+	config "istio.io/api/mixer/v1/config/descriptor"
+)
+
+// Func every expression function provider should have this interface.
+type Func interface {
+	// Name uniquely identified the function.
+	Name() string
+
+	// ReturnType specifies the return type of this function.
+	ReturnType() config.ValueType
+
+	// ArgTypes specifies the argument types in order expected by the function.
+	ArgTypes() []config.ValueType
+
+	// NullArgs specifies if the function accepts null args
+	NullArgs() bool
+
+	// Call performs the function call. It is guaranteed
+	// that call will be made with correct types and arity.
+	// may panic.
+	Call([]interface{}) interface{}
+}
+
+type baseFunc struct {
+	name     string
+	argTypes []config.ValueType
+	retType  config.ValueType
+	nullArgs bool
+}
+
+func (f *baseFunc) Name() string                 { return f.name }
+func (f *baseFunc) ReturnType() config.ValueType { return f.retType }
+func (f *baseFunc) ArgTypes() []config.ValueType { return f.argTypes }
+func (f *baseFunc) NullArgs() bool               { return f.nullArgs }
+
+type eqFunc struct {
+	*baseFunc
+	invert bool
+}
+
+// newEQFunc returns dynamically typed equality fn.
+// treats strings specially with prefix and suffix '*' matches
+func newEQFunc() Func {
+	return &eqFunc{
+		baseFunc: &baseFunc{
+			name:     "EQ",
+			retType:  config.BOOL,
+			argTypes: []config.ValueType{config.VALUE_TYPE_UNSPECIFIED, config.VALUE_TYPE_UNSPECIFIED},
+		},
+	}
+}
+
+// newNEQFunc returns inverse of newEQFunc fn.
+func newNEQFunc() Func {
+	return &eqFunc{
+		baseFunc: &baseFunc{
+			name:     "NEQ",
+			retType:  config.BOOL,
+			argTypes: []config.ValueType{config.VALUE_TYPE_UNSPECIFIED, config.VALUE_TYPE_UNSPECIFIED},
+		},
+		invert: true,
+	}
+}
+
+func (f *eqFunc) Call(args []interface{}) interface{} {
+	res := f.call(args)
+	if f.invert {
+		return !res
+	}
+	return res
+}
+
+func (f *eqFunc) call(args []interface{}) bool {
+	var s0 string
+	var s1 string
+	var ok bool
+
+	if s0, ok = args[0].(string); !ok {
+		return args[0] == args[1]
```

Comment:
This doesn't work for cases where args are things like `[]byte`; I think you might need to use [`reflect.DeepEqual`](https://golang.org/pkg/reflect/#DeepEqual) here.